src/allmydata/test/test_magic_folder.py

```diff
@@ -1518,6 +1525,64 @@ def test_delete(self):
         self.assertTrue(node is not None, "Failed to find %r in DMD" % (path,))
         self.failUnlessEqual(metadata['version'], 1)
 
+    @defer.inlineCallbacks
+    def test_batched_process(self):
+        # setup: get at least two items into the deque
+        path0 = os.path.join(self.local_dir, u'foo')
+        yield self.fileops.write(path0, 'foo\n')
+        path1 = os.path.join(self.local_dir, u'bar')
+        yield self.fileops.write(path1, 'bar\n')
+
+        # get the status before we've processed anything
+        upstatus0 = list(self.magicfolder.uploader.get_status())
+        upstatus1 = []
+
+        def one_item(item):
+            # grab status after we've processed a single item
+            us = list(self.magicfolder.uploader.get_status())
+            upstatus1.extend(us)
+        one_d = self.magicfolder.uploader.set_hook('item_processed')
+        # can't 'yield' here because the hook isn't called until
+        # inside iterate()
+        one_d.addCallbacks(one_item, self.fail)
+
+        yield iterate_uploader(self.magicfolder)
+        yield iterate_uploader(self.magicfolder)  # req'd for windows; not sure why?
+
+        # status we got each time should be the same
+        self.assertEqual(len(upstatus0), len(upstatus1))
+        for item0, item1 in zip(upstatus0, upstatus1):
+            self.assertEqual(item0, item1)
+
+    @defer.inlineCallbacks
+    def test_real_notify_failure(self):
+
+        def bad_stuff(*args, **kw):
+            # the function we replaced would trigger this hook, so we
+            # have to or things will stall
+            self.magicfolder.uploader._call_hook(u"foo", "inotify")
+            raise RuntimeError("the bad stuff")
+
+        patch_notify = mock.patch.object(
+            self.magicfolder.uploader,
+            '_real_notify',
+            mock.Mock(side_effect=bad_stuff),
+        )
+        with patch_notify:
+            path0 = os.path.join(self.local_dir, u'foo')
+            yield self.fileops.write(path0, 'foo\n')
+
+        # do a reactor turn; this is necessary because our "bad_stuff"
+        # method calls the hook (so the above 'yield' resumes) right
+        # *before* it raises the exception; thus, we ensure all the
+        # pending callbacks including the exception are processed
+        # before we flush the errors.
+        yield task.deferLater(reactor, 0, lambda: None)
+        errors = self.flushLoggedErrors()
```

Comment:
You could also pass `RuntimeError` to `flushLoggedErrors` to restrict which failures get flushed from the log.

My favorite pattern for this is to define a new exception type inside the test method that can't be raised by anything else.  Then flush that exception type.  Then the length assertion is good enough.  If you got one flushed item then it was the thing you wanted to get logged.

Reply:
Ah, nice I like this pattern suggestion :) :+1: